File: board.cpp
#include "board.h"
#include <iostream>
#include <QTextStream>
#include <QVector>


#include "logger.h"
#include "wordcollector.h"
#include "cell.h"


Board::Board(QObject *parent) :
    QObject(parent)
{
    isChanged = false;
    isApproved = false;

    currentPlayer = FIRST_PLAYER;

}

void Board::initBoard(int width, int height){
    WIDTH = width;
    HEIGHT = height;
    for (int i = 0; i < HEIGHT; ++i){
        for(int j = 0; j < WIDTH; ++j) {
            board_.push_back(QVector<Cell*>());
            board_[i].push_back(new Cell(QChar('-')));
        }
    }
}


//Connections

void Board::setUpConnection(WordCollector* wordCollector) {
    connect(this, &Board::commitLetter, wordCollector, &WordCollector::addLetter);
    //connect(this, SIGNAL(commitLetter(QChar)), wordCollector, SLOT(addLetter(QChar)));
    connect(this, SIGNAL(commitX(int)), wordCollector, SLOT(addX(int)));
    connect(this, SIGNAL(commitY(int)), wordCollector, SLOT(addY(int)));
    connect(this, SIGNAL(commitWord()), wordCollector, SLOT(checkWord()));
    connect(this, SIGNAL(commitNew(int)), wordCollector, SLOT(addNew(int)));
    connect(this, SIGNAL(addNewLetter(QPair<int,int>)), wordCollector, SLOT(addChangedCell(QPair<int,int>)));
}

void Board::connectToPlayers(QObject* player1, QObject* player2) {
    connect(this, SIGNAL(moveEnded(QString)), player1, SLOT(approveWord(QString)));
    connect(this, SIGNAL(chooseError(QString)), player1, SLOT(badChooseLetter(QString)));
    connect(this, SIGNAL(letterChosen(QPair<QPair<int,int>, QChar>)), player1, SLOT(letterChosen(QPair<QPair<int,int>, QChar >)));
    connect(this, SIGNAL(sendBoardFirst(QVector<QVector<QChar> >)), player1, SLOT(setCurrentBoard(QVector<QVector<QChar> >)));
    connect(this, SIGNAL(resetWordFirst(const QPair<int,int>&)),
            player1, SLOT(onBoardResetWord(const QPair<int,int>&)));

    connect(this, SIGNAL(moveEndedSecond(QString)), player2, SLOT(approveWord(QString)));
    connect(this, SIGNAL(chooseErrorSecond(QString)), player2, SLOT(badChooseLetter(QString)));
    connect(this, SIGNAL(letterChosenSecond(QPair<QPair<int,int>, QChar>)), player2, SLOT(letterChosen(QPair<QPair<int,int>, QChar>)));
    connect(this, SIGNAL(sendBoardSecond(QVector<QVector<QChar> >)), player2, SLOT(setCurrentBoard(QVector<QVector<QChar> >)));
    connect(this, SIGNAL(resetWordSecond(const QPair<int,int>&)),
            player2, SLOT(onBoardResetWord(const QPair<int,int>&)));

}

void Board::connectToGameManager(QObject* gameManager) {
    connect(this, SIGNAL(sendCellsNumber(int)), gameManager, SLOT(getNumberOfCells(int)));
}


//Methods

void Board::setFirstWord(QString firstWord) {
    for (int i = 0; i <  WIDTH; ++i) {
        board_[HEIGHT/2][i]->setLetter(firstWord[i]);
    }
}

void Board::setFirstPlayer(int player) {
    currentPlayer = player;
}
void Board::sendError(QString message) {
    if (currentPlayer == FIRST_PLAYER) {
        emit chooseError(message);
    } else {
        emit chooseErrorSecond(message);
    }
}

bool Board::rangeCheck(int x, int y) {
    if (x < 0 || x >= WIDTH || y < 0 || y >= HEIGHT) {
        sendError(tr("Wrong range of cell"));
        return false;
    }
    return true;
}

void Board::showBoard(){
    QTextStream out(stdout);
    for(int i = 0; i < HEIGHT; ++i){
        for(int j = 0; j < WIDTH; ++j) {
            out << board_[i][j]->getLetter() << " ";
        }
        out <<"\n";
    }
}

void Board::changeLetter(int x, int y, QChar letter) {
    if (hasChanged()) {
        sendError(tr("Cell has already choosen"));
        return;
    }
    if (!rangeCheck(x, y)) {
        return;
    }
    if (getLetter(x, y) != QChar('-')) {
        sendError(tr("Error cell chosen"));
        return;
    }

    board_[x][y]->setLetter(letter);
    isChanged = true;
    setMarked(x, y, true);
    setChanged(true);
    showBoard();
    QPair<QPair<int,int>, QChar> sendLetter = qMakePair(qMakePair(x, y), letter);
    emit addNewLetter(QPair<int,int>(x, y));
    if (currentPlayer == FIRST_PLAYER) {
        emit letterChosen(sendLetter);
    } else {
        emit letterChosenSecond(sendLetter);
    }

}

void Board::pushLetter(QPair<int, int>& coordinates) {
    if (!rangeCheck(coordinates.first, coordinates.second)) {
        return;
    }
    if (getLetter(coordinates.first, coordinates.second) == QChar('-')) {
        return;
    }
    emit commitLetter(getLetter(coordinates.first, coordinates.second));
    emit commitX(coordinates.first);
    emit commitY(coordinates.second);
    if (isMarked(coordinates.first, coordinates.second)) {
        emit commitNew(1);
    } else {
        emit commitNew(0);
    }
}

bool Board::isMarked(int x, int y) {
    return board_[x][y]->isMarked();
}

void Board::setMarked(int x, int y, bool value) {
    board_[x][y]->setMarked(value);
}

QChar Board::getLetter(int x, int y) {
    return board_[x][y]->getLetter();
}

void Board::setLetter(int x, int y, QChar letter) {
    board_[x][y]->setLetter(letter);
}

bool Board::hasChanged() {
    return isChanged;
}

void Board::setChanged(bool changed) {
    isChanged = changed;
}



//slots

void Board::resetState(const QPair<int,int>& coordinates) {
    isChanged = false;
    for (int i = 0; i < WIDTH; ++i) {
        for (int j = 0; j < HEIGHT; ++j) {
            if (getLetter(i,j) != '-') {
                setMarked(i, j, false);
            }
        }
    }
    std::cout<<"reset state from" << coordinates.first << " " << coordinates.second << std::endl;
    std::cout.flush();
    setLetter(coordinates.first, coordinates.second, '-');
    setMarked(coordinates.first, coordinates.second, false);
    if (currentPlayer == FIRST_PLAYER) {
        emit resetWordFirst(coordinates);
    } else {
        emit resetWordSecond(coordinates);
    }
}

void Board::remakeMove(const QString& word) {

    std::cout<<"New move\n";
    std::cout << std::endl;
    isChanged = false;
    for (int i = 0; i < WIDTH; ++i) {
        for (int j = 0; j < HEIGHT; ++j) {
            if (getLetter(i,j) != '-') {
                setMarked(i, j, false);
            }
        }
    }
    if (isApproved) {
        //switchPlayer;
        isApproved = false;
        currentPlayer = 3 - currentPlayer;
        switch (currentPlayer) {
        case 2:
            emit moveEnded(word);
            break;

        case 1:
            emit moveEndedSecond(word);
            break;
        }
    } else {
        showBoard();
    }
}

void Board::setApproved() {
    isApproved = true;
}

/*
 * Slots from Player
 * */

void Board::chooseLetterFirst(QPair<QPair<int,int>, QChar>& letter) {
    changeLetter(letter.first.first, letter.first.second, letter.second);
    Logger l;
    l.printLog<QString>(DEBUG, tr("GET"));
    l.printLog<int>(DEBUG, letter.first.first);
    l.printLog<int>(DEBUG, letter.first.second);
    l.printLog<QChar>(DEBUG, letter.second);
}

void Board::chooseLetterSecond(QPair<QPair<int,int>,QChar>& letter) {
    changeLetter(letter.first.first, letter.first.second, letter.second);
}

void Board::pushLetterFirst(QPair<int, int>& coordinates) {
    pushLetter(coordinates);
}
void Board::pushLetterSecond(QPair<int, int>& coordinates) {
    pushLetter(coordinates);
}

void Board::showBoardToPlayer() {
    QVector<QVector<QChar> > result(WIDTH, QVector<QChar>(HEIGHT));
    for (int i = 0; i < WIDTH; ++i) {
        for (int j = 0; j < HEIGHT; ++j) {
            result[i][j] = getLetter(i,j);
        }
    }
    if (currentPlayer == FIRST_PLAYER) {
        emit sendBoardFirst(result);
    } else {
        emit sendBoardSecond(result);
    }

}


void Board::gotCommitQuery() {
   emit commitWord();
}

void Board::getNumberOfCells() {
    int cnt = 0;
    for (int i = 0; i < WIDTH; ++i) {
        for (int j = 0; j < HEIGHT; ++j) {
            if (getLetter(i, j) == QChar('-')) {
                ++cnt;
            }
        }
    }
    emit sendCellsNumber(cnt);
}

void Board::showBoardToManager() {
    showBoard();
}

File: python-src/Board.py
import unittest
from Cell import Cell
from Letter import CellLetter, Coordinates
from PySide import QtCore
from WordCollector import WordCollector

__author__ = 'user1'


class Board(QtCore.QObject):

    #signals to WordCollector


    commit_letter = QtCore.Signal(str)
    commit_x = QtCore.Signal(int)
    commit_y = QtCore.Signal(int)
    commit_word = QtCore.Signal()
    commit_new = QtCore.Signal(int)
    add_new_letter = QtCore.Signal(CellLetter)
    move_ended = QtCore.Signal(str)
    choose_error = QtCore.Signal(str)
    letter_chosen = QtCore.Signal(tuple)
    send_board_first = QtCore.Signal(list)
    reset_word_first = QtCore.Signal(Coordinates)

    move_ended_second = QtCore.Signal(str)
    choose_error_second = QtCore.Signal(str)
    letter_chosen_second = QtCore.Signal(tuple)
    send_board_second = QtCore.Signal(list)
    reset_word_second = QtCore.Signal(Coordinates)

    """signals to GameManager"""

    send_cells_number = QtCore.Signal(int)


    def __init__(self):
        super(Board, self).__init__()
        self.__board__ = list()
        self.FIRST_PLAYER = 1
        self.SECOND_PLAYER = 2
        self.__current_player__ = self.FIRST_PLAYER
        self.__is_changed__ = False
        self.__is_approved__ = False

        self.WIDTH = 5
        self.HEIGHT = 5



    """Slots from WordCollector"""

    @QtCore.Slot(Coordinates)
    def reset_state(self, coordinates : Coordinates):
        self.__is_changed__ = False
        for i in range(self.WIDTH):
            for j in range(self.HEIGHT):
                if self.get_letter(i, j) != '-':
                    self.set_marked(i, j, False)
        print("Reset state from", coordinates.x, coordinates.y)
        self.set_letter(coordinates.x, coordinates.y, '-')
        self.set_marked(coordinates.x, coordinates.y, False)
        if self.__current_player__ == self.FIRST_PLAYER:
            self.reset_word_first.emit(coordinates)
        else:
            self.reset_word_second.emit(coordinates)

    @QtCore.Slot(str)
    def remake_move(self, word):
        print("New move")
        self.__is_changed__ = False
        for i in range(self.WIDTH):
            for j in range(self.HEIGHT):
                if self.get_letter(i, j) != '-':
                    self.set_marked(i, j, False)
        if self.__is_approved__:
            self.__is_approved__ = False
            self.__current_player__ = 3 - self.__current_player__
            if self.__current_player__ == 2:
                self.move_ended.emit(word)
            else:
                self.move_ended_second.emit(word)

    @QtCore.Slot()
    def set_approved(self):
        self.__is_approved__ = True

    """Slots from Player"""

    @QtCore.Slot(CellLetter)
    def choose_letter_first(self, letter : CellLetter):
        self.change_letter(letter.x, letter.y, letter.letter)
        print("GET")


    @QtCore.Slot(CellLetter)
    def choose_letter_second(self, letter: CellLetter):
        self.change_letter(letter.x, letter.y, letter.letter)

    @QtCore.Slot()
    def show_board_to_player(self):
        result = []
        for i in range(self.HEIGHT):
            #TODO Inspect for index order
            result.append([])
            for j in range(self.WIDTH):
                result[i].append(self.get_letter(i, j))
        if self.__current_player__ == self.FIRST_PLAYER:
            self.send_board_first.emit(result)
        else:
            self.send_board_second.emit(result)

    @QtCore.Slot(Coordinates)
    def push_letter_first(self, coordinates: Coordinates):
        self.push_letter(coordinates)

    @QtCore.Slot(Coordinates)
    def push_letter_second(self, coordinates: Coordinates):
        self.push_letter(coordinates)

    @QtCore.Slot()
    def got_commit_query(self):
        self.commit_word.emit()

    """Slots from GameManager"""

    @QtCore.Slot()
    def get_number_of_cells(self):
        cnt = 0
        for i in range(self.WIDTH):
            for j in range(self.HEIGHT):
                if self.get_letter(i, j) == '-':
                    cnt += 1
        self.send_cells_number.emit(cnt)

    @QtCore.Slot()
    def show_board_to_manager(self):
        self.show_board()

    """Public methods"""
    def send_error(self, message):
        if self.__current_player__ == self.FIRST_PLAYER:
            self.choose_error.emit(message)
        else:
            self.choose_error_second.emit(message)

    def range_check(self, x, y):
        if x < 0 or x >= self.WIDTH or y < 0 or y >= self.HEIGHT:
            self.send_error("Wrong range of cell")
            return False
        return True

    def set_first_player(self, player):
        self.__current_player__ = player

    def set_first_word(self, first_word: str):
        for i in range(self.HEIGHT):
            self.__board__[self.WIDTH // 2][i].set_letter(first_word[i])

    def setup_connection(self, word_collector: WordCollector):
        self.commit_letter.connect(word_collector.add_letter)
        self.commit_x.connect(word_collector.add_x)
        self.commit_y.connect(word_collector.add_y)
        self.commit_word.connect(word_collector.check_word)
        self.commit_new.connect(word_collector.add_new)
        self.add_new_letter.connect(word_collector.add_changed_cell)

    def connect_to_players(self, player1, player2):
        self.move_ended.connect(player1.approve_word)
        self.choose_error.connect(player1.bad_choose_letter)
        self.letter_chosen.connect(player1.letter_chosen)
        self.send_board_first.connect(player1.set_current_board)
        self.reset_word_first.connect(player1.on_board_reset_word)

        self.move_ended_second.connect(player2.approve_word)
        self.choose_error.connect(player2.bad_choose_letter)
        self.letter_chosen_second.connect(player2.letter_chosen)
        self.send_board_second.connect(player2.set_current_board)
        self.reset_word_second.connect(player2.on_board_reset_word)

    def connect_to_game_manager(self, game_manager):
        self.send_cells_number.connect(game_manager.get_number_of_cells)

    def show_board(self):
        cnt = 0
        for index_row in range(self.WIDTH):
            for index_cell in range(self.HEIGHT):
                print(self.__board__[index_row][index_cell].get_letter(), end=" ")
            print()


    def change_letter(self, x, y, letter):
        if self.has_changed():
            self.send_error("Cell has already chosen")
            return
        if not self.range_check(x, y):
            return
        if self.get_letter(x, y) != '-':
            self.send_error("Error cell chosen")
            return
        self.__board__[x][y].set_letter(letter)
        self.__is_changed__  = True
        self.set_marked(x, y, True)
        self.show_board()
        send_letter = CellLetter(x, y, letter)
        self.add_new_letter.emit(send_letter)
        if self.__current_player__ == self.FIRST_PLAYER:
            self.letter_chosen.emit(send_letter)
        else:
            self.letter_chosen_second.emit(send_letter)

    def push_letter(self, coordinates: Coordinates):
        if not self.range_check(coordinates.x, coordinates.y):
            return
        if self.get_letter(coordinates.x, coordinates.y) == '-':
            return
        self.commit_letter.emit(self.get_letter(coordinates.x, coordinates.y))
        self.commit_x.emit(coordinates.x)
        self.commit_y.emit(coordinates.y)
        if self.is_marked(coordinates.x, coordinates.y):
            self.commit_new.emit(1)
        else:
            self.commit_new.emit(0)

    def is_marked(self, x, y):
        return self.__board__[x][y].is_marked()

    def set_marked(self, x, y, value):
        self.__board__[x][y].set_marked(value)

    def get_letter(self, x, y):
        return self.__board__[x][y].get_letter()

    def set_letter(self, x, y, letter):
        self.__board__[x][y].set_letter(letter)

    def has_changed(self):
        return self.__is_changed__

    def init_board(self, width, height):
        self.WIDTH = width
        self.HEIGHT = height
        for i in range(width):
            for j in range(height):
                self.__board__.append([])
                self.__board__[i].append(Cell('-'))


def setup_connection(board : Board, word_collector: WordCollector):
    board.commit_letter.connect(word_collector.add_letter)
    board.commit_x.connect(word_collector.add_x)
    board.commit_y.connect(word_collector.add_y)
    board.commit_word.connect(word_collector.check_word)
    board.commit_new.connect(word_collector.add_new)
    board.add_new_letter.connect(word_collector.add_changed_cell)


class BoardConnectionTest(unittest.TestCase):
    def setUp(self):
        self.board = Board()
        self.wc = WordCollector()
        self.board.init_board(5,5)
        self.board.set_first_word('carry')

    def test_connect(self):
        setup_connection(self.board, self.wc)



if __name__ == '__main__':
    unittest.main()